### PR TITLE
Adding a function that could deal with the magnetic_force and spin when using the ABACUS's deltaspin module

### PR DESCRIPTION
**Pull request by flinky-ai**
## Adding a new function, two files were changed.

I add three parameters in the data when reading data from the OUT.ABACUS/running_scf.log.

When we use the deltaspin module during the scf calculation, the dpdata could grep the magnetic force and spin to the system and print the modified coords.npy and forces.npy.

## Summary by CodeRabbit

- **Bug Fixes**
  - Corrected import statement for `py3Dmol` to ensure proper functionality.

- **Refactor**
  - Reorganized and reformatted import statements across multiple plugins for improved code readability.
  - Adjusted method signatures and added new functions to enhance the functionality of specific formats.

- **Chores**
  - Updated `.pre-commit-config.yaml` with new URLs and revisions for repositories.
  - Modified `.gitattributes` to enforce line endings for Python test files.

### CODE REVIEW

**Comment by flinky-ai** [comment]
@amcadmus @chazeon I cloned the original dpdata code into my local space. But failed to pip install with these codes. Now, I add some function into the dpdata to deal with the deltaspin_scf abacus calculation results. And now, it failed with the PR process. I checked my code. It should be workable. Could you please help me with it?
```
Defaulting to user installation because normal site-packages is not writeable
Processing /mnt/d/AISI/dpdata
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
Building wheels for collected packages: UNKNOWN
  Building wheel for UNKNOWN (pyproject.toml) ... done
  Created wheel for UNKNOWN: filename=UNKNOWN-0.2.19.dev20+ge1f02fd.d20240515-py3-none-any.whl size=4091 sha256=85056635321a8add11d86137b7469a82fa56dd77b985682e5def93568b00e9df
  Stored in directory: /tmp/pip-ephem-wheel-cache-9kbvnwiy/wheels/00/6a/08/3bb0e14df3403c9233763f1104d7951eb1cd83c31155a6d7b0
Successfully built UNKNOWN
Installing collected packages: UNKNOWN
  Attempting uninstall: UNKNOWN
    Found existing installation: UNKNOWN 0.1.dev698+g658a511
    Uninstalling UNKNOWN-0.1.dev698+g658a511:
      Successfully uninstalled UNKNOWN-0.1.dev698+g658a511
Successfully installed UNKNOWN-0.2.19.dev20+ge1f02fd.d20240515
```

**Comment by chazeon** [comment]
The failed CI build seems to come from readthedocs, due to Python 3.12:
* [AttributeError: module 'pkgutil' has no attribute 'ImpImporter'. Did you mean: 'zipimporter'?](https://stackoverflow.com/questions/77364550/attributeerror-module-pkgutil-has-no-attribute-impimporter-did-you-mean)

The readthedocs build environment is recently updated; see #655.
You could try to rebase to the `devel` branch first, then change the PR target to `devel`.

For your local install, you need to clean up your workspace; it seems to be reading the package info from the `PKG-INFO` you created in your earlier commits.

Finally, I would suggest you include a test example with spins in your PR.

File: dpdata/abacus/scf.py [hunk 1]
```diff
@@ -157,6 +157,44 @@ def get_energy(outlines):
...
+    outlines (list of str) : The lines from the output file.
+    natoms (int) : The number of atoms.
+
+    Returns
+    -------
+    np.ndarray: An array of magnetic moments.
+    """
+    Mag = []
+
+    # 定义正则表达式以匹配所需的数字
```
Comment (njzjz) [inline dpdata/abacus/scf.py:175]: We don't accept any non-English content.

File: dpdata/abacus/scf.py [hunk 2]
```diff
@@ -157,6 +157,44 @@ def get_energy(outlines):
...
+    Extracts magnetic moments from the output lines.
+
+    Parameters
+    ----------
+    outlines (list of str) : The lines from the output file.
+    natoms (int) : The number of atoms.
+
+    Returns
+    -------
+    np.ndarray: An array of magnetic moments.
```
Comment (njzjz) [inline dpdata/abacus/scf.py:171]: The doc does not meet the [standard](https://numpydoc.readthedocs.io/en/latest/format.html).

File: dpdata/abacus/scf.py
Comment (njzjz) [inline dpdata/abacus/scf.py:1]: Unit tests should be added.

File: dpdata/abacus/scf.py [hunk 3]
```diff
@@ -280,10 +411,36 @@ def get_frame(fname):
...
+        mag_forces = None
+        coords_deltaspin = None
+        force_deltaspin = None
+
+    # print(force_deltaspin)
+    # print(coords_deltaspin)
+    # print(spin)
+    # print(mag_forces)
+    # print(force)
+    # print(coords)
```
Comment (njzjz) [inline dpdata/abacus/scf.py:431]: Unused codes should be removed.

File: dpdata/abacus/scf.py [hunk 4]
```diff
@@ -280,10 +411,36 @@ def get_frame(fname):
...
     data["cells"] = cell[np.newaxis, :, :]
     data["coords"] = coords[np.newaxis, :, :]
     data["energies"] = np.array(energy)[np.newaxis]
     data["forces"] = force[np.newaxis, :, :]
+    if deltaspin is not None:
+        data["mag_forces"] = mag_forces[np.newaxis, :, :]
+        data["spin"] = spin[np.newaxis, :, :]
+        data["coords_deltaspin"] = coords_deltaspin[np.newaxis, :, :]
+        data["force_deltaspin"] = force_deltaspin[np.newaxis, :, :]
+        data["deltaspin"] = deltaspin
```
Comment (njzjz) [inline dpdata/abacus/scf.py:442]: New data types should be registered in `System.DTYPES`.

File: dpdata/abacus/scf.py [hunk 5]
```diff
@@ -230,6 +327,34 @@ def get_stress(outlines):
...
+            for line in file:
+                # 移除行首尾的空白字符，然后以空格分割
+                parts = line.strip().split()
+                # 检查是否是sc_mag_switch参数行
+                if len(parts) >= 2 and parts[0] == "sc_mag_switch":
+                    # 检查参数值是否为1
+                    return True if parts[1] == "1" else None
+        # 文件已读完，没有找到sc_mag_switch参数，返回None
+        return None
+    except FileNotFoundError:
```
Comment (njzjz) [inline dpdata/abacus/scf.py:353]: This is never touched, considering it has been opened in line 384. Besides, it should not be read twice.

**Review by njzjz — CHANGES_REQUESTED** [review]

File: dpdata/abacus/scf.py
Comment (flinky-ai) [reply dpdata/abacus/scf.py:1]: How should I add the Unit test to my code? I didn't find any documents or codes about 'Unit test'.

**Comment by flinky-ai** [comment]
> The failed CI build seems to come from readthedocs, due to Python 3.12:
> 
> * [AttributeError: module 'pkgutil' has no attribute 'ImpImporter'. Did you mean: 'zipimporter'?](https://stackoverflow.com/questions/77364550/attributeerror-module-pkgutil-has-no-attribute-impimporter-did-you-mean)
> 
> The readthedocs build environment is recently updated; see #655. You could try to rebase to the `devel` branch first, then change the PR target to `devel`.
> 
> For your local install, you need to clean up your workspace; it seems to be reading the package info from the `PKG-INFO` you created in your earlier commits.
> 
> Finally, I would suggest you include a test example with spins in your PR.

I used the `pip -vvv install .` and found the details of my installation while install the cloned dpdata locally. And it seems to be the reason that my `setuptools` version is not replaced by the newly installed `setputools-69.5.1`. The problem is solved by upgrading `setuptools` manually.


```
Successfully installed packaging-24.0 setuptools-69.5.1 setuptools_scm-8.1.0 tomli-2.0.1
  Installing build dependencies ... done
  Running command Getting requirements to build wheel
  /tmp/pip-build-env-nvbne_99/overlay/local/lib/python3.10/dist-packages/setuptools_scm/_integration/setuptools.py:31: RuntimeWarning:
  ERROR: setuptools==59.6.0 is used in combination with setuptools_scm>=8.x

  Your build configuration is incomplete and previously worked by accident!
  setuptools_scm requires setuptools>=61
```

```
:/mnt/d/AISI/dpdata$ pip3 install --upgrade pip setuptools
Defaulting to user installation because normal site-packages is not writeable
Requirement already satisfied: pip in /usr/lib/python3/dist-packages (22.0.2)
Collecting pip
  Downloading pip-24.0-py3-none-any.whl (2.1 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.1/2.1 MB 15.1 kB/s eta 0:00:00
Requirement already satisfied: setuptools in /usr/lib/python3/dist-packages (59.6.0)
Collecting setuptools
  Using cached setuptools-69.5.1-py3-none-any.whl (894 kB)
Installing collected packages: setuptools, pip
Successfully installed pip-24.0 setuptools-69.5.1
```

**Review by njzjz — CHANGES_REQUESTED** [review]
It looks like you changed the line ending of each file from LF to CRLF, which should be reverted

**Comment by flinky-ai** [comment]
> It looks like you changed the line ending of each file from LF to CRLF, which should be reverted

I don't know in what step led me to make this mistake. I will change the format and resubmit again.

File: dpdata/abacus/scf.py [hunk 6]
```diff
@@ -297,6 +394,10 @@ def get_frame(fname):
     # print("energy = ", data['energies'])
     # print("force = ", data['forces'])
     # print("virial = ", data['virials'])
+    # print("spin = ", data['spin'])
+    # print("mag_forces = ", data['mag_forces'])
+    # print("force_deltaspin = ", data['force_deltaspin'])
+    # print("coords_deltaspin = ", data['coords_deltaspin'])
```
Comment (njzjz) [inline dpdata/abacus/scf.py:400]: commented codes should be removed

**Review by njzjz — CHANGES_REQUESTED** [review]

**Pull request closed without merging**